### ISSUE TITLE
Feature Request : Upload Progress

### ISSUE THREAD

**Issue by calumk** [pocketbase/pocketbase#2364]
I know this is referenced in a discussion on the Dart SDK already, but dont think there is a specific issue for it.

At the moment the lack of file upload progress is quite problematic - espeically for larger files 100mb+

Is this on the road-map?

**Comment by ganigeorgiev** [comment]
This is in my internal todo, but for now is a very low priority and at least for the near future I'm not planning working on it. For JS in the browser, you can manually send plain XHR request and use its [`upload` prop](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/upload). For the Dart SDK the issue is slightly more complicated since it depends how the file was loaded.

I'll keep the issue open for easier tracking, but if anyone want to pick it feel free to outline some of the planned implementation details (_I believe this'll require changes mostly in the SDKs_).

**Comment by calumk** [comment]
Thanks for the reply 

For anyone looking to do this, I managed to get it to work (for multiple simultanious file uploads)

Look at the Records API, to see how to patch a record: https://pocketbase.io/docs/api-records/
i used nanoid() to generate a unique ID for each upload, to track the progress of each file.

Obviously sample below is just an exmaple, but might be helpful to others.

```javascript

let filesArr = []

for (var i = 0; i < e.target.files.length; i++) {
        const formData = new FormData();
        formData.append("files", e.target.files[i])

        let _uploadID = nanoid()
        filesArr.push({_uploadID : _uploadID, name : e.target.files[i].name, progress : 0})

        let xhr = new XMLHttpRequest();
        xhr.upload._uploadID = _uploadID

        // listen for upload progress
        xhr.upload.onprogress = function(event) {
            let percent = Math.round(100 * event.loaded / event.total);
            filesArr.find(item => item._uploadID === this._uploadID).progress = percent
            console.log(`File ${this._uploadID} is ${percent} % uploaded...`)
        };

        // handle error
        xhr.upload.onerror = function() {
            console.log(`Error during the upload: ${xhr.status}.`);
        };

        // upload completed successfully
        xhr.onload = async function(event) {
            console.log('Upload completed successfully.');
        };

        xhr.open('PATCH', 'http://pocketbaseurl.com/api/collections/collectionIdOrName/records/recordId');
        xhr.setRequestHeader("Authorization", pb.authStore.token);
        xhr.send(formData);
        
    }
```

**Comment by NickMundel** [comment]
I tried the same with axios and it seems that I always only receive two respones one being when the file upload is at 50 bytes and later when it has finished. Have you found a work-around for this?

**Comment by fcastrovilli** [comment]
i'm trying to screw this but xhr req works only in the browser so i can't use it in server side code like other crud processes (would love to use it in sveltekit as form action). this means i have to directly expose authstore token to client side for auth headers and i don't really like the idea (am i paranoid?). anyway totally agree with others saying that is so crucial to have feedbacks with medium files uploads. are you planning to increase this feature priority in the near future?

**Comment by RayyanNafees** [comment]
> I tried the same with axios and it seems that I always only receive two responses one being when the file upload is at 50 bytes and later when it has finished. Have you found a work-around for this?

Have u tried with axios' [onUploadProgress](https://stackoverflow.com/a/41088162) config ?

**Comment by ganigeorgiev** [comment]
_Old issues cleanup: I'm closing this as I don't plan working on it. Users contributions to the respective SDKs are welcomed._

**Comment by calumk** [comment]
Update for anyone who stumbles upon this.

I have reworked my original example a few times recently.
I found that running multiple patch commands simultaneously could cause issues where the files were not correctly uploaded.
It seems to be important to let one patch command finish before starting the next. So this is now "sequential"

```js

let nanoid = () => {
    return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
}

let filesArr = []

let uploadFiles = async (e) => {
    filesArr = [];

    for (var i = 0; i < e.target.files.length; i++) {
        const formData = new FormData();
        formData.append("FILES_COLUMN_NAME_CHANGE_ME+", e.target.files[i]);

        let _uploadID = nanoid();
        filesArr.push({
            _uploadID: _uploadID,
            name: e.target.files[i].name,
            progress: 0,
        });

        let xhr = new XMLHttpRequest();
        xhr.upload._uploadID = _uploadID;
        xhr.upload._filename = e.target.files[i].name;

        // Promisify the XMLHttpRequest
        await new Promise((resolve, reject) => {
            // listen for upload start
            xhr.upload.onloadstart = function () {
                console.log('Upload started.', this._uploadID, this._filename);
            };

            // listen for upload progress
            xhr.upload.onprogress = function (event) {
                let percent = Math.round((100 * event.loaded) / event.total);
                filesArr.find(item => item._uploadID === this._uploadID).progress = parseInt(percent);
            };

            // handle error
            xhr.upload.onerror = function () {
                console.log(`Error during the upload: ${xhr.status}.`, this._uploadID, this._filename);
                reject(new Error(`Upload failed with status: ${xhr.status}`));
            };

            // upload completed successfully
            xhr.upload.onload = async function () {
                console.log('Upload completed successfully.', this._uploadID, this._filename);
                filesArr = filesArr.filter(item => item._uploadID !== this._uploadID);
                await getFilesFromCourse();
                resolve();
            };

            xhr.open('PATCH', `https://${pocketbaseurl}/api/collections/${collectionId}/records/${recordId}`);
            xhr.setRequestHeader("Authorization", pb.authStore.token);
            xhr.send(formData);
        });
    }
};

```
